### PR TITLE
storage: fix local_call not finding C stored procedures

### PR DESCRIPTION
**Pull request by Serpentian**
Currently, if function was created as C stored procedure or as Lua
persistent function (with body argument) via box.schema.func.create,
all types of router.call and router.map_callrw cannot find it and
return `Procedure 'name' is not defined` error.

This is cased by the fact that both of these function use local_call,
which invokes net_box.self.call. It didn't work with these type of
functions before Tarantool 3.0.0-beta1-18. Let's use box.func, where
it's needed instead of net_box.self.call in local_call.

Closes https://github.com/tarantool/vshard/issues/436

NO_DOC=bugfix

### CODE REVIEW

File: vshard/storage/init.lua
Comment (Gerold103) [inline vshard/storage/init.lua:251]: > NO_TEST=<already tested>

But where?

File: vshard/storage/init.lua
Comment (Gerold103) [inline vshard/storage/init.lua:258]: If something is working one way for all versions < N and in another way for all versions > N, then this doesn't have to be a runtime check. Please, make `local_call()` depend on the core version. For earlier versions we can try the fallback approach. For newer ones - simply always use `netbox_self_call()` like before.

As "make it depend on core version" I mean that you can define `local_call()` differently depending on the version. Similar to how `util.index_min()` is defined.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: vshard/storage/init.lua
Comment (Serpentian) [reply vshard/storage/init.lua:266]: It was in commit msg, but github suppresses everything inside <...> in PR's comment. However, this is not important now, as I added the test, as it turns out this problem also happens with functions, created with `box.schema.func.create(..., {body = " ... "})`

File: vshard/storage/init.lua
Comment (Serpentian) [reply vshard/storage/init.lua]: > If something is working one way for all versions < N and in another way for all versions > N, then this doesn't have to be a runtime check.

We've already done version dependent thing via runtime checks: [netbox_is_conn_dead](https://github.com/tarantool/vshard/blob/9fda3546ded3f74fa03dda371a984986e2b367d1/vshard/replicaset.lua#L165C1-L185).

We cannot use `netbox_self_call` for newer versions, as there's no version, where it works now. I don't know, when I'll do it ([ticket](https://github.com/tarantool/tarantool/issues/9131)). So, let's use runtime check for now.

Comment (Serpentian) [reply vshard/storage/init.lua]: The only version dependent code we can make now: before 2.2.0 always use `netbox_self_call`, as there're no `box.func` API, in all other cases try to fallback. I don't really see much sense in such approach, but this can be done, if you insist

Comment (Gerold103) [reply vshard/storage/init.lua]: > netbox_is_conn_dead

Here it is justified, because it has the fast-path which is version-agnostic and works almost always:
```Lua
    if conn.state == 'active' then
        return false
    end
```
In your case you've added branching on the path which is executed always, so this is different.

> We cannot use netbox_self_call for newer versions, as there's no version, where it works now

If this is not fixed even in master, then we can't merge nor test it until it is done in master. If this is already done in master, then you can always check even a not yet released version by the commit number. For example, the latest master I see is `3.0.0-beta1-5-ge844cd09b`, which means 5 commits after `3.0.0-beta1`. `vshard.version` module supports commit numbers, so you can do "if version >= 3.0.0-beta1-5 then use the new way, otherwise use the fallback approach".

Comment (Serpentian) [reply vshard/storage/init.lua]: No, it's not done in master, the bug is still there (it's not related to release). Nonetheless, I don't get, why we're blocking the PR, if we can fix it in vshard right now, it's even tested.

So, at the end of a day we mark https://github.com/tarantool/vshard/issues/436 as blocked by https://github.com/tarantool/tarantool/issues/9131?

File: vshard/storage/init.lua
Comment (Gerold103) [inline vshard/storage/init.lua:278]: Lets at least avoid the double-lookup in `box.func` and do it like this:
```Lua
local func = box.func and box.func[func_name]
if not func then
    return pcall(netbox_self_call, netbox_self, func_name, args)
end
return pcall(func.call, func, args)
```

File: vshard/storage/init.lua
Comment (Gerold103) [reply vshard/storage/init.lua]: Ok, I see the full picture now, thanks for the explanation. Lets finalize this PR and file a ticket to make the `local_call` depend on the version in the future.

Comment (Serpentian) [reply vshard/storage/init.lua]: https://github.com/tarantool/tarantool/issues/9131 was fixed by the Core team. Let's make version banching

**Review by Gerold103 — APPROVED** [review]
Thanks for the fixes!

**Pull request merged**
